[PECO-239] Apache Arrow support

kravets-levko opened the pull request:
[PECO-239] Add Apache Arrow support

- [x] Basic Apache arrow support
- [x] `TSparkArrowTypes` support
- [x] Unit tests
- [x] E2E tests
- [x] Have Arrow feature always enabled
- [x] Option to control if native arrow types should be used or not
- [x] Return Arrow result as regular javascript object (convert from internal `apache-arrow` classes)
  - [x] Should be compatible with existing JSON result handler
  - [x] If native types disabled - should parse JSON-serialized fields similarly to existing JSON result handler
  - [x] Tests

Arrow support is always enabled. User can choose whether to represent complex types as JSON-encoded strings or as native Arrow objects by setting a `options.useArrowNativeTypes` option of `IDBSQLSession.executeStatement` (defaults to `true`)

`TSparkArrowTypes.intervalTypesAsArrow` is not currently supported by `apache-arrow` (see https://github.com/streamlit/streamlit/issues/4489) and therefore not exposed to user. PySQL also doesn't support this option, so at least we're consistent here

[PECO-239]: https://databricks.atlassian.net/browse/PECO-239?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

andrefurlan-db commented:
Before merging this I'd like to review the interfaces and tests with @rcypher-databricks and Go driver.

andrefurlan-db commented on lib/DBSQLSession.ts:
```diff
@@ -42,6 +48,41 @@ function getDirectResultsOptions(maxRows: number | null = defaultMaxRows) {
+  const useArrowNativeTypes: TSparkArrowTypes = {
+    timestampAsArrow: arrowOptions?.useNativeTimestamps,
+    decimalAsArrow: arrowOptions?.useNativeDecimals,
+    complexTypesAsArrow: arrowOptions?.useNativeComplexTypes,
```
if we know that this returns invalid strings, why do we allow users to change? What is the use case of enabling arrow and disabling some data types? We should not expose flags to users that we know have problems

kravets-levko replied:
As we discussed, I removed most of the options, made Arrow always enabled, and replaced native types options with a single one - to completely enable/disable the usage of Arrow native types. For now it may be useful for a case if user encounter some bug in our Arrow response processing code, but once we're confident that it works flawlessly - we could deprecate/remove this option as well

andrefurlan-db commented on tests/e2e/batched_fetch.test.js:
```diff
@@ -27,14 +27,14 @@ describe('Data fetching', () => {
   it('fetch chunks should return a max row set of chunkSize', async () => {
     const session = await openSession();
-    const operation = await session.executeStatement(query, { runAsync: true, maxRows: null });
+    const operation = await session.executeStatement(query, { runAsync: true, maxRows: null, enableArrow: false });
```
we should add tests to ensure that the fetching logic still works for arrow batches

andrefurlan-db commented on tests/e2e/arrow.test.js:
```diff
@@ -0,0 +1,310 @@
+          int_type: 4000000,
+          flt: 1.2000000476837158,
+          dbl: 2.2,
+          dec: new arrow.util.BN(new Uint32Array([320, 0, 0, 0])), // This field is affected
```
wait, I thought the desired behaviour would be to return native types. Arrow should be default and it should not impact the results...

andrefurlan-db commented on tests/e2e/arrow.test.js:
```diff
@@ -0,0 +1,310 @@
+          vchr: 'b',
+          dat: new Date('2014-01-17T00:00:00.000Z'),
+          arr_type: '[2014-01-17,2023-01-22]',
+          map_type: '{"key":1991-01-01}',
```
I also thought we had agreed that allowing users to set a flag that returns results that cannot be parsed is bad...

andrefurlan-db commented on tests/e2e/batched_fetch.test.js:
```diff
@@ -27,14 +27,14 @@ describe('Data fetching', () => {
   it('fetch all should fetch all records', async () => {
     const session = await openSession();
-    const operation = await session.executeStatement(query, { runAsync: true, maxRows: null });
+    const operation = await session.executeStatement(query, { runAsync: true, maxRows: null, enableArrow: false });
```
why do you have to disable arrow for this test to pass? I'd expect that arrow or not would not impact this.

andrefurlan-db reviewed (commented):
a few surprises here.
Missing: 
tests that confirms that arrow on or off do not change the result.
tests to ensure that pagination logic still works with arrow.

kravets-levko commented:
@andrefurlan-db I updated tests so they cover most data types handling. Also, added a separate test to make sure that reponse with multiple arrow batches is also handled as it should. Please take a look. Thank you!

andrefurlan-db commented on lib/result/ArrowResult.ts:
```diff
@@ -0,0 +1,137 @@
+    }
+    if (value instanceof MapRow) {
+      const result = value.toJSON();
+      // Map type consists of its key and value types. We need only value type here, key will be cast to string anyway
```
are you sure? I thought `map[any]any` was the type definition. So you could have a struct as key, no?

kravets-levko replied:
this is for compatibility with column results - maps are json-serialized as plain objects which could have only string keys. we could try to use JS built-in `Map` class to represent maps, but for column results it probably will have some limitations

kravets-levko replied:
I mentioned it in Readme in tests/fixtures folder. In short: complex types used as map keys produce malformed json, so they could be only handled with Arrow + native types enabled. And even in this case, since JS objects cannot have complex type keys, they will remain stringified. So it will look something like this:

```js
{ 
  '{"s": "hello"}': ..., 
  '{"s": "world"}': ..., 
}
```

andrefurlan-db commented on lib/result/ArrowResult.ts:
```diff
@@ -0,0 +1,137 @@
+    if (value instanceof Object && value[isArrowBigNumSymbol]) {
+      const result = bignumToBigInt(value);
+      if (DataType.isDecimal(valueType)) {
+        return Number(result) / 10 ** valueType.scale;
```
are you sure this is correct? float, double, and decimal have different representations and conversion between them may result in differences. If the difference is lost due to arrow, the nevermind my comment.

kravets-levko replied:
here the issue actually is that JS has no any native fixed-point data type, and you cannot introduce your own that will look like a native type. tere is the only numeric type in JS which is simultaneously integer and float (depending on assigned value). in JSON-serialized response, `decimal` type is represented as float, Arrow represents it as integer. so here I basically convert that integer to JS number

andrefurlan-db replied:
👍

andrefurlan-db commented on lib/result/ArrowResult.ts:
```diff
@@ -0,0 +1,137 @@
+      return result.map((item) => this.convertArrowTypes(item, field?.type, field?.type.children || []));
+    }
+
+    if (DataType.isTimestamp(valueType)) {
```
what about Date?

kravets-levko replied:
Date is already converted internally

andrefurlan-db commented on lib/result/utils.ts:
```diff
@@ -0,0 +1,80 @@
+}
+
+function convertDate(value: any): Date {
+  return value instanceof Date ? value : new Date(Date.parse(`${value} UTC`));
```
depending on the timezone of the session, the date returned may not be in UTC.

andrefurlan-db replied:
also, anything needs to be done to handle negative dates?

kravets-levko replied:
good catch, thank you! I'll double-check this

andrefurlan-db replied:
👍

andrefurlan-db commented on lib/DBSQLSession.ts:
```diff
@@ -42,6 +49,32 @@ function getDirectResultsOptions(maxRows: number | null = defaultMaxRows) {
 
+function getArrowOptions(useArrowNativeTypes: boolean | undefined): {
+  canReadArrowResult: boolean;
+  useArrowNativeTypes?: TSparkArrowTypes;
```
Didn't we decide that we should not expose this flag to users? Why would our user not use arrow native types?

kravets-levko replied:
Sorry, missed this coment. Addressed in https://github.com/databricks/databricks-sql-nodejs/pull/94/commits/2b8a0d6f2c17ab9cdced0db8e4c7d2c84b4711fc

andrefurlan-db commented on tests/fixtures/compatibility/index.js:
```diff
@@ -0,0 +1,24 @@
+    // point numbers are double-precision, so after decoding we get, for example,
+    // 1.414199948310852 instead of 1.4142. Therefore, we need to "fix" value
+    // returned by Arrow, so it can be tested against our fixture
+    flt: Number(row.flt.toFixed(4)),
```
uh, this is unfortunate. I am curious to know what alternative "fixes" for this problem you tried

kravets-levko replied:
I think there's no good fix. Basically, that's not even an issue itself, just add some extra frictions in tests, so we don't need to fix it in our code anyhow. You can see the same behavior in any language that has separate `single` and `double` types - when you cast `single` to `double` and the value is expanded. The same happens here - Arrow has separate `single` and `double` types, but JS has only Number which internally is `double`. So IMHO this "fix" - to compare values using only as much decimals as we specified in sample data - is good enough

andrefurlan-db commented on tests/fixtures/compatibility/README.md:
```diff
@@ -0,0 +1,14 @@
+
+Known issues:
+
+- with Arrow disabled _or_ with Arrow native types disabled:
```
@rcypher-databricks , what is the other serialization issue with arrow disabled?

kravets-levko replied:
those cases are the ones I stumbled at. there's even a thread in Slack about date serialization

andrefurlan-db reviewed (commented):
Almost there! Added a few comments.

kravets-levko commented on lib/globalConfig.ts:
```diff
@@ -0,0 +1,3 @@
+export default {
+  arrowEnabled: true,
```
a sort of. currently I needed it for test, but probably later will refine this for a wider use

andrefurlan-db reviewed (approved):
looks good

susodapop reviewed (approved):
Stamp. I updated the PAT for e2e tests and kicked-off a new execution.
